### PR TITLE
Deployment

### LINKED ISSUE #3: Remove `zserver/dev` script
I assume this script should build the server into a bundle that can be run directly via `node`.

Here's the full log:

```
npm run zserver/dev

> hollowverse@1.0.0 zserver/dev D:\code\hollowverse\hollowverse
> webpack --config server/webpack.dev.js

ts-loader: Using typescript@2.2.2 and D:\code\hollowverse\hollowverse\tsconfig.json
Hash: 6dab2407d091a14653e4
Version: webpack 2.3.2
Time: 6046ms
    Asset     Size  Chunks                    Chunk Names
bundle.js  1.91 MB       0  [emitted]  [big]  main
   [0] ./~/node-libs-browser/~/process/browser.js 5.3 kB {0} [built]
  [14] ./~/path-browserify/index.js 6.18 kB {0} [built]
  [31] ./~/react/lib/React.js 2.69 kB {0} [built]
  [48] ./~/react/react.js 56 bytes {0} [built]
 [178] ./~/redux/es/index.js 1.08 kB {0} [built]
 [183] ./client/src/main.app.tsx 350 bytes {0} [built] [failed] [1 error]
 [184] ./~/express/index.js 224 bytes {0} [built]
 [185] ./~/react-dom/server.js 65 bytes {0} [built]
 [186] ./~/react-redux/lib/index.js 475 bytes {0} [built]
 [187] ./server/src/redux.store.ts 404 bytes {0} [built]
 [188] ./client/src/redux.reducers.ts 389 bytes {0} [built] [failed] [1 error]
 [242] ./~/express/lib/express.js 2.22 kB {0} [built]
 [340] ./~/react-dom/lib/ReactDOMServer.js 735 bytes {0} [built]
 [408] ./~/redux-thunk/lib/index.js 529 bytes {0} [built]
 [429] ./server/src/index.tsx 1.9 kB {0} [built]
    + 428 hidden modules

WARNING in ./~/express/lib/view.js
80:29-41 Critical dependency: the request of a dependency is an expression

ERROR in ./client/src/main.app.tsx
Module parse failed: D:\code\hollowverse\hollowverse\client\src\main.app.tsx The keyword 'interface' is reserved (12:0)
You may need an appropriate loader to handle this file type.
| import {Link} from 'react-router'
|
| interface Props {
|   loginStatus: facebookSdk.LoginStatus,
|   isNavMenuOpen: boolean
 @ ./server/src/index.tsx 6:19-55

ERROR in ./client/src/redux.reducers.ts
Module parse failed: D:\code\hollowverse\hollowverse\client\src\redux.reducers.ts Unexpected token (6:19)
You may need an appropriate loader to handle this file type.
| import {IGeneralState, initialGeneralState, actionHandlers} from './redux.actions'
|
| import LoginStatus = facebookSdk.LoginStatus
|
| interface IState extends IGeneralState {
 @ ./server/src/redux.store.ts 4:25-67
 @ ./server/src/index.tsx

ERROR in ./~/express/lib/request.js
Module not found: Error: Can't resolve 'net' in 'D:\code\hollowverse\hollowverse\node_modules\express\lib'
 @ ./~/express/lib/request.js 18:11-25
 @ ./~/express/lib/express.js
 @ ./~/express/index.js
 @ ./server/src/index.tsx

ERROR in ./~/express/lib/view.js
Module not found: Error: Can't resolve 'fs' in 'D:\code\hollowverse\hollowverse\node_modules\express\lib'
 @ ./~/express/lib/view.js 18:9-22
 @ ./~/express/lib/application.js
 @ ./~/express/lib/express.js
 @ ./~/express/index.js
 @ ./server/src/index.tsx

ERROR in ./~/send/index.js
Module not found: Error: Can't resolve 'fs' in 'D:\code\hollowverse\hollowverse\node_modules\send'
 @ ./~/send/index.js 24:9-22
 @ ./~/express/lib/response.js
 @ ./~/express/lib/express.js
 @ ./~/express/index.js
 @ ./server/src/index.tsx

ERROR in ./~/etag/index.js
Module not found: Error: Can't resolve 'fs' in 'D:\code\hollowverse\hollowverse\node_modules\etag'
 @ ./~/etag/index.js 22:12-25
 @ ./~/express/lib/utils.js
 @ ./~/express/lib/application.js
 @ ./~/express/lib/express.js
 @ ./~/express/index.js
 @ ./server/src/index.tsx

ERROR in ./~/destroy/index.js
Module not found: Error: Can't resolve 'fs' in 'D:\code\hollowverse\hollowverse\node_modules\destroy'
 @ ./~/destroy/index.js 14:17-30
 @ ./~/send/index.js
 @ ./~/express/lib/response.js
 @ ./~/express/lib/express.js
 @ ./~/express/index.js
 @ ./server/src/index.tsx

ERROR in ./~/mime/mime.js
Module not found: Error: Can't resolve 'fs' in 'D:\code\hollowverse\hollowverse\node_modules\mime'
 @ ./~/mime/mime.js 2:9-22
 @ ./~/send/index.js
 @ ./~/express/lib/response.js
 @ ./~/express/lib/express.js
 @ ./~/express/index.js
 @ ./server/src/index.tsx

npm ERR! Windows_NT 10.0.15063
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "zserver/dev"
npm ERR! node v6.6.0
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE
npm ERR! hollowverse@1.0.0 zserver/dev: `webpack --config server/webpack.dev.js`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the hollowverse@1.0.0 zserver/dev script 'webpack --config server/webpack.dev.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the hollowverse package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     webpack --config server/webpack.dev.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs hollowverse
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls hollowverse
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\code\hollowverse\hollowverse\npm-debug.log
```

### LINKED ISSUE #5: `npm run client/build` fails
This may be an issue with my setup. I'll check.

@msafih, do all scripts (`npm run ...`) work after a clean `npm install` for you?

```
npm run client/build

> hollowverse@1.0.0 client/build D:\code\hollowverse\hollowverse
> webpack --optimize-minimize --define process.env.NODE_ENV="'production'" --config client/webpack.dev.js

ts-loader: Using typescript@2.2.2 and D:\code\hollowverse\hollowverse\tsconfig.json
Hash: 81e9293dee7edb8f2fb4
Version: webpack 2.3.2
Time: 12526ms
             Asset      Size  Chunks                    Chunk Names
         bundle.js   2.07 MB       0  [emitted]  [big]  main
    css/styles.css   94.7 kB       0  [emitted]         main
     bundle.js.map   2.53 MB       0  [emitted]         main
css/styles.css.map  91 bytes       0  [emitted]         main
   [3] ./~/react/react.js 56 bytes {0} [built]
  [42] ./~/react-redux/lib/index.js 475 bytes {0} [built]
  [52] ./~/react-router-redux/lib/index.js 1.97 kB {0} [built]
  [84] ./~/react-router/es/index.js 1.46 kB {0} [built]
 [104] ./~/react-dom/index.js 59 bytes {0} [built]
 [156] ./client/src/styles.index.scss 41 bytes {0} [built]
 [157] ./client/src/main.index.tsx 1.2 kB {0} [built]
 [293] ./~/react-router/es/IndexLink.js 597 bytes {0} [built]
 [297] ./~/react-router/es/Router.js 5.31 kB {0} [built]
 [299] ./~/react-router/es/applyRouterMiddleware.js 1.9 kB {0} [built]
 [341] ./client/src/main.app.tsx 3.67 kB {0} [built]
 [342] ./client/src/page.createProfile.tsx 1.59 kB {0} [built]
 [343] ./client/src/page.homepage.tsx 3.76 kB {0} [built]
 [347] ./client/src/redux.store.ts 920 bytes {0} [built]
 [352] ./client/src/webpackEntry.ts 195 bytes {0} [built]
    + 339 hidden modules

ERROR in bundle.js from UglifyJs
Invalid assignment [./client/src/utils.ts:6,27][bundle.js:3639,31]
Child extract-text-webpack-plugin:
       [0] ./~/css-loader/lib/css-base.js 1.46 kB {0} [built]
       [1] ./~/css-loader!./~/sass-loader/lib/loader.js!./client/src/styles.index.scss 94.8 kB {0} [built]

npm ERR! Windows_NT 10.0.15063
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "client/build"
npm ERR! node v6.6.0
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE
npm ERR! hollowverse@1.0.0 client/build: `webpack --optimize-minimize --define process.env.NODE_ENV="'production'" --config client/webpack.dev.js`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the hollowverse@1.0.0 client/build script 'webpack --optimize-minimize --define process.env.NODE_ENV="'production'" --config client/webpack.dev.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the hollowverse package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     webpack --optimize-minimize --define process.env.NODE_ENV="'production'" --config client/webpack.dev.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs hollowverse
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls hollowverse
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     D:\code\hollowverse\hollowverse\npm-debug.log
```

### LINKED ISSUE #17: Set up a live instance
We need to do two things to make life better:

1. Deploy what we currently have - [apparently may be as easy as `firebase deploy`](https://firebase.google.com/docs/hosting/quickstart).
2. Set up Continuous Deployment - maybe [with Travis CI](https://docs.travis-ci.com/user/deployment/firebase/) after passing tests from `mater`.

This issue exists to remind about the first step.

### LINKED ISSUE #23: Compile TypeScript ES6 to ES5 with Babel to fix production build
Currently, we're unable to build with webpack using the parameter `--optimize-minimize` because it uses UglifyJS which fails on ES6. So we need to pipe TypeScript output to Babel to compile down to ES5 so that we can minify for production.

### PR DESCRIPTION
**Pull request by wholesomedev**
I have initted a Firebase project at the root of our repo. Basically, I followed the Firebase hosting instructions (https://firebase.google.com/docs/hosting/) and that asked me to install the Firebase CLI (https://firebase.google.com/docs/cli/) and init a project. I did that and here's the structure that they gave us:

* `/public` is gonna contain all the static assets (JS, CSS, HTML, images, etc)
* `/functions` is gonna contain [Firebase Cloud Functions](https://firebase.google.com/docs/functions/)

(There are also other dot files and configuration files that they added.)

### Domain name

My plan is to create the domain name **`x.hollowverse.com`** and have that be our home for now. Then I want to create a deploy script. 

### Deploy script

The job of the deploy script is the following:

1. Create the JavaScript bundle and put it in `/public`
1. Create `styles.css` and put it in `/public`
1. Copy `index.html` from `/client` to `/public`
1. Run `firebase deploy` (that will upload `/public` and `/functions` to Firebase)

The command to run the deploy script will be **`npm run deploy`**.

I'm thinking of using ShellJS to perform the tasks of the deploy script. I'd like to avoid using Grunt or Gulp if possible.

### Other notes

For now we don't have any Cloud Functions, but I think it should be part of our design guidelines that any code that can be run on the server, should be run on the server. For example, we have code in `vendor.firebase.ts` that handles creation of new user that can be moved server-side. We should move it. I've added that bit to our README.

It should also be noted that the deployment process will change before we launch **v1**. We will need an actual Node server to do server-side rendering and we'll need automatic deployment of `master` through Travis CI...and at that point we'll probably wanna remove `npm run deploy`. And eventually we'll want to use a chat bot to do deployment of branches other than `master`.

### Remaining tasks

- [x] Setup domain name
- [x] Create a build script
- [x] Add deployment instructions to README

### Fixes

Fixes #23, fixes #17, fixes #5, fixes #3

### CODE REVIEW

**Comment by wholesomedev** [comment]
I've setup the domain name. `https://x.hollowverse.com`. But I haven't deployed any content to it yet.

**Comment by deltaidea** [comment]
I've searched a bit and it looks like you're right, there's no way to do server-side rendering without a dedicated Node.JS server.

I'd propose using Heroku since it's so easy to set up, has a great free plan with unlimited apps for experiments, and I have experience with it. We can switch to a less hipster hosting once there's a noticeable load from users.

**Review by deltaidea — CHANGES_REQUESTED** [review]
Lemme configure Travis CI, Heroku, do a full deploy and then we can merge with confidence.

File: README.md [hunk 1]
```diff
@@ -22,10 +22,17 @@ This will provide a dev version of the front-end on `localhost:8080` and rebuild
...
 
-* Start the server via nodemon: `npm run server/dev`
+* Start the server via nodemon: `npm run server/express`
 
 This will run the server straight from the source in `./server/src`. Nodemon will automatically refresh it on changes.
 
+### Building in production mode and deploying
+
+* Build assets in production mode and run Firebase Hosting locally: `npm run server/firebase`
+* Deploy to Firebase: `npm run deploy` - requires access to the Firebase project
```
Comment (wholesomedev) [inline README.md]: I think in addition to having access to the Firebase project, they also need to have run `firebase login` in order for `npm run deploy` to work.

**Comment by wholesomedev** [comment]
@deltaidea Can we do the Node.js, server-side rendering, and Travis CI setup as part of other PRs?

Regarding Heroku, I think it is a good choice, but since we're already on Google Cloud platform, I propose we stay with it. I'm not familiar with Google Cloud platform myself, but eventually we'll have to get familiar with it, so we might as well do it now. Things will be simpler if we do not add another service provider to the mix.

**Comment by wholesomedev** [comment]
So I think with your latest commits, this PR is ready to review and merge.

**Review by deltaidea — APPROVED** [review]

**Comment by deltaidea** [comment]
Agreed on all points. 👍 
Check the last two commits and go ahead and merge this bad boy.

**Pull request merged**